workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/SpatialOSBehaviourManager.cs

```diff
@@ -1,60 +1,221 @@
+using System;
 using System.Collections.Generic;
+using Improbable.Gdk.Core.MonoBehaviours;
 using Improbable.Worker;
 using UnityEngine;
+using Entity = Unity.Entities.Entity;
 
 namespace Improbable.Gdk.Core
 {
-    public class ReaderWriterImpl
-    {
-        // This is placeholder code.
-    }
-
     /// <summary>
-    ///     Invokes dispatcher callbacks on GameObjects that represent entities.
+    ///     Keeps track of Reader/Writer availability for SpatialOSBehaviours on a particular GameObject and decides when
+    ///     a SpatialOSBehaviour should be enabled, calling into the SpatialOSBehaviourLibrary for injection.
     /// </summary>
-    public class SpatialOSBehaviourManager
+    internal class SpatialOSBehaviourManager
     {
-        public GameObject GameObject;
-        private Dictionary<uint, List<MonoBehaviour>> componentIdToReaderSpatialOSBehaviours;
-        private Dictionary<uint, List<MonoBehaviour>> componentIdToWriterSpatialOSBehaviours;
-        private Dictionary<MonoBehaviour, int> numUnsatisfiedComponents;
-        private List<MonoBehaviour> spatialOSBehavioursToEnable;
-        private List<MonoBehaviour> spatialOSBehavioursToDisable;
-        private Dictionary<uint, Dictionary<MonoBehaviour, ReaderWriterImpl>> readersWriters;
+        private readonly Entity entity;
+        private readonly long spatialId;
+
+        private readonly Dictionary<uint, HashSet<MonoBehaviour>> behavioursRequiringReaderTypes
+            = new Dictionary<uint, HashSet<MonoBehaviour>>();
+
+        private readonly Dictionary<uint, HashSet<MonoBehaviour>> behavioursRequiringWriterTypes
+            = new Dictionary<uint, HashSet<MonoBehaviour>>();
+
+        private readonly Dictionary<MonoBehaviour, int> numUnsatisfiedReadersOrWriters
+            = new Dictionary<MonoBehaviour, int>();
+
+        private readonly Dictionary<MonoBehaviour, Dictionary<uint, IReaderInternal>> behaviourToReadersWriters
+            = new Dictionary<MonoBehaviour, Dictionary<uint, IReaderInternal>>();
+
+        private readonly Dictionary<uint, HashSet<IReaderInternal>> compIdToReadersWriters =
+            new Dictionary<uint, HashSet<IReaderInternal>>();
+
+        private readonly HashSet<MonoBehaviour> behavioursToEnable = new HashSet<MonoBehaviour>();
+        private readonly HashSet<MonoBehaviour> behavioursToDisable = new HashSet<MonoBehaviour>();
 
-        public SpatialOSBehaviourManager(GameObject gameObject)
+        private readonly SpatialOSBehaviourLibrary behaviourLibrary;
+
+        private readonly ILogDispatcher logger;
+
+        private const string LoggerName = "SpatialOSBehaviourManager";
```

Comment:
I'm personally in favour of getting the loggername programmatically, i.e. using `this.GetType().ToString()` or something similar rather than saving a constant.

Reply:
That would be nice, but I don't see a way to do that without making it a non-static field, and I'd rather not store additional data on each object without a good reason.

Reply:
My idea was that rather than declaring a const string here, we can just get the loggername every time we want to send a log, i.e. ```
logger.HandleLog(LogType.Error, new LogEvent(MultipleReadersWritersRequiredError).WithField(LoggingUtils.LoggerName, GetType());
```